### PR TITLE
Provefixes

### PR DESCRIPTION
**Pull request by MrChico**
Adds missing `not` and `GEQ` to parsing and type checking.
Refactors `Prove` to use a single context.
Removes strange storage bound post condition for constructor specs.
Make post conditions a proof obligation instead of assumption for `act prove`.
Introduces some CPS tricks in Main procedure.

### CODE REVIEW

File: src/Prove.hs [hunk 1]
```diff
@@ -395,10 +397,23 @@ locsFromExp e = case e of
...
+
+
+toInt :: SMType -> Maybe (SBV Integer)
+toInt (SymInteger i) = Just i
+toInt _ = Nothing
+
+toBool :: SMType -> Maybe (SBV Integer)
+toBool (SymInteger i) = Just i
+toBool _ = Nothing
+
```
Comment (MrChico) [inline src/Prove.hs:412]:
```suggestion
```

File: src/Prove.hs [hunk 2]
```diff
@@ -398,7 +400,7 @@ get name vars = fromMaybe (error (name <> " not found in " <> show vars)) $ Map.
 catInts :: Map Id SMType -> Map Id (SBV Integer)
 catInts m = Map.fromList $ go $ Map.toList m
   where
-    go ((name, SymInteger i):tl) = (name, i):(go tl)
+    go ((name, (SymInteger i)):tl) = (name, i):(go tl)
```
Comment (MrChico) [inline src/Prove.hs:403]:
```suggestion
    go ((name, SymInteger i):tl) = (name, i):(go tl)
```

File: src/Prove.hs [hunk 3]
```diff
@@ -179,135 +177,139 @@ mkStorageConstraints preCtx@(Ctx _ m _ preStore _ pre) (Ctx _ _ _ postStore _ po
     mkConstraint (Left loc) = fromLocation loc
     mkConstraint (Right update) = fromUpdate update
 
+    getVar :: (Show b) => TStorageItem a -> (Map Id (SMType, SMType) -> Map Id b) -> b
+    getVar i f = get (nameFromItem m i) (f store)
```
Comment (xwvvvvwx) [inline src/Prove.hs:181]:
```suggestion
    getVar i f = get (nameFromItem m i) (f store)
    
```

File: src/Prove.hs [hunk 4]
```diff
@@ -54,12 +54,12 @@ queries claims = fmap mkQueries gathered
 type Contract = Id
 type Method = Id
 type Args = Map Id SMType
-type Store = Map Id SMType
+type Store = Map Id (SMType, SMType)
 type Env = Map Id SMType
 data When = Pre | Post
-  deriving (Eq, Show)
+ deriving (Eq, Show)
```
Comment (xwvvvvwx) [inline src/Prove.hs:60]:
```suggestion
  deriving (Eq, Show)
```

**Review by xwvvvvwx — APPROVED** [review]

File: src/Prove.hs [hunk 5]
```diff
@@ -2,7 +2,7 @@
 {-# LANGUAGE DataKinds #-}
 {-# LANGUAGE OverloadedStrings #-}
 
-module Prove (queries) where
+module Prove (queries, Ctx(..)) where
```
[the remark this answers is not in the document]
Comment (MrChico) [reply src/Prove.hs:5]: I will reuse it in the hevm backend

**Pull request merged**